### PR TITLE
enh(api): make acknowledgement endpoints consistent with downtimes

### PR DESCRIPTION
**Pull request by kduret**
## Description

Acknowledgment endpoints was not consistent with downtime endpoints

**Fixes** MON-4638

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Check new endpoints in api documentation and validate these

### CODE REVIEW

File: doc/API/centreon-api-v2.yaml [hunk 1]
```diff
@@ -473,26 +583,27 @@ paths:
           $ref: '#/components/responses/InternalServerError'
     post:
       tags:
-        - Acknowledgment
-      summary: "Add an acknowledgment on host"
+        - Acknowledgement
+      summary: "Add an acknowledgement on host"
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]:
```suggestion
      summary: "Add an acknowledgement on selected host"
```

File: doc/API/centreon-api-v2.yaml [hunk 2]
```diff
@@ -473,26 +583,27 @@ paths:
...
           $ref: '#/components/responses/Forbidden'
         '500':
           $ref: '#/components/responses/InternalServerError'
-  /acknowledgments/hosts/{host_id}:
     delete:
       tags:
-        - Acknowledgment
-      summary: "Cancel an acknowledgment on host"
+        - Acknowledgement
+      summary: "Cancel an acknowledgement on host"
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]:
```suggestion
      summary: "Cancel an acknowledgement on a host"
```

File: doc/API/centreon-api-v2.yaml [hunk 3]
```diff
@@ -502,16 +613,18 @@ paths:
           $ref: '#/components/responses/Forbidden'
         '500':
           $ref: '#/components/responses/InternalServerError'
-  /acknowledgments/services:
+  /monitoring/hosts/{host_id}/services/{service_id}/acknowledgements:
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]: shouldn't the API be like : /acknowledgements/hosts/:id/services/:id

File: doc/API/centreon-api-v2.yaml [hunk 4]
```diff
@@ -533,26 +646,28 @@ paths:
           $ref: '#/components/responses/InternalServerError'
     post:
       tags:
-        - Acknowledgment
-      summary: "Add an acknowledgment on service"
+        - Acknowledgement
+      summary: "Add an acknowledgement on service"
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]:
```suggestion
      summary: "Add an acknowledgement on a service"
```

File: doc/API/centreon-api-v2.yaml [hunk 5]
```diff
@@ -533,26 +646,28 @@ paths:
...
           $ref: '#/components/responses/Forbidden'
         '500':
           $ref: '#/components/responses/InternalServerError'
-  /acknowledgments/hosts/{host_id}/services/{service_id}:
     delete:
       tags:
-        - Acknowledgment
-      summary: "Cancel an acknowledgment on service"
+        - Acknowledgement
+      summary: "Cancel an acknowledgement on service"
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]:
```suggestion
      summary: "Cancel an acknowledgement on a service"
```

File: doc/API/centreon-api-v2.yaml [hunk 6]
```diff
@@ -1265,50 +1389,38 @@ components:
...
-    Acknowledgment.Host.Add:
+    Acknowledgement.Host.Add:
       type: object
       properties:
         comment:
           type: string
-          description: "Short description of the acknowledgment"
+          description: "Short description of the acknowledgement"
           example: "Acknowledged by admin"
-        host_id:
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]: shouldn't the hostId be specified ?

File: doc/API/centreon-api-v2.yaml [hunk 7]
```diff
@@ -1265,50 +1389,38 @@ components:
...
-          description: "Indicates whether acknowledgment is maintained in the case of a change of status"
+          description: "Indicates whether acknowledgement is maintained in the case of a change of status"
           example: true
-    Acknowledgment.Service.Add:
+    Acknowledgement.Service.Add:
       allOf:
-        - $ref: '#/components/schemas/Acknowledgment.Host.Add'
-        - type: object
-          properties:
-            service_id:
```
Comment (sc979) [inline doc/API/centreon-api-v2.yaml]: shouldn't the serviceId be specified ?

File: src/Centreon/Domain/Acknowledgement/Acknowledgement.php [hunk 8]
```diff
@@ -104,18 +111,18 @@ class Acknowledgement
     private $isSticky;
 
     /**
-     * @Serializer\Groups({"ack_service"})
+     * @Serializer\Groups({"ack_main"})*
```
Comment (sc979) [inline src/Centreon/Domain/Acknowledgement/Acknowledgement.php]: why use snake_case ?

File: src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php [hunk 9]
```diff
@@ -0,0 +1,574 @@
+<?php
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+        if ($this->hasNotEnoughRightsToContinue()) {
+            return null;
+        }
+
+        // Internal call for non admin user
+        return $this->findOneAcknowledgement($acknowledgementId, false);
+    }
+
+    /**
+     * Find one acknowledgement tacking into account or not the ACLs.
```
Comment (sc979) [inline src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php]:
```suggestion
     * Find one acknowledgement taking into account or not the ACLs.
```

File: src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+    {
+        if ($this->hasNotEnoughRightsToContinue()) {
+            return null;
+        }
+
+        $accessGroupFilter = $this->isAdmin()
+            ? ' '
+            : ' INNER JOIN `:dbstg`.`centreon_acl` acl
+                  ON acl.host_id = ack2.host_id
+                INNER JOIN `:db`.`acl_groups` acg
```
Comment (sc979) [inline src/Centreon/Infrastructure/Acknowledgement/AcknowledgementRepositoryRDB.php]: shouldn't ':db' be bound too before the execute step ?

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+            $this->engineService,
+            $this->entityValidator
+        );
+
+        $this->expectException(ValidationFailedException::class);
+        $acknowledgementService->addServiceAcknowledgement($this->serviceAcknowledgement);
+    }
+
+    /**
+     * test addServiceAcknowledgement whith not found service
```
Comment (sc979) [inline tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]:
```suggestion
     * test addServiceAcknowledgement with not found service
```

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+            $this->engineService,
+            $this->entityValidator
+        );
+
+        $this->expectException(EntityNotFoundException::class);
+        $acknowledgementService->addServiceAcknowledgement($this->serviceAcknowledgement);
+    }
+
+    /**
+     * test addServiceAcknowledgement whith not found host
```
Comment (sc979) [inline tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]:
```suggestion
     * test addServiceAcknowledgement with not found host
```

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+            $this->engineService,
+            $this->entityValidator
+        );
+
+        $this->expectException(ValidationFailedException::class);
+        $acknowledgementService->addHostAcknowledgement($this->serviceAcknowledgement);
+    }
+
+    /**
+     * test addHostAcknowledgement whith not found host
```
Comment (sc979) [inline tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]:
```suggestion
     * test addHostAcknowledgement with not found host
```

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+            $this->entityValidator
+        );
+        $acknowledgementService->filterByContact($this->aclContact);
+
+        $acknowledgement = $acknowledgementService->findAcknowledgements();
+        $this->assertEquals($acknowledgement, [$this->serviceAcknowledgement]);
+    }
+
+    /**
+     * test addHostAcknowledgement which is not validated
```
Comment (sc979) [inline tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]: do you mean not processed by engine ?

File: doc/API/centreon-api-v2.yaml (hunk 6 above)
Comment (kduret) [reply doc/API/centreon-api-v2.yaml]: it is specified in the URI now : /monitoring/hosts/{hostId}/acknowledgements

File: doc/API/centreon-api-v2.yaml (hunk 7 above)
Comment (kduret) [reply doc/API/centreon-api-v2.yaml]: it is specified in the URI now : /monitoring/hosts/{hostId}/services/{serviceId}/acknowledgements

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php (hunk 15 above)
Comment (kduret) [reply tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]: Not really
This test that an exception is thrown if the host entity is not well formated (eg: host id is not set)

File: src/Centreon/Domain/Acknowledgement/Acknowledgement.php (hunk 8 above)
Comment (kduret) [reply src/Centreon/Domain/Acknowledgement/Acknowledgement.php]: Honnestly don't know exctly why. But may be related to the yaml convention

File: tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php (hunk 15 above)
Comment (sc979) [inline tests/php/Centreon/Domain/Acknowledgement/AcknowledgementServiceTest.php]:
```suggestion
     * test addHostAcknowledgement which is not valid (eg: missing hostId)
```

**Pull request merged**